hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapFetchIndexOperation.java

```diff
@@ -0,0 +1,281 @@
+                        return migratedPartitionIds.contains(partitionId);
+                    })
+                    .collect(Collectors.toList());
+        }
+
+        response = new MapFetchIndexOperationResult(entries, migratedPartitionIds, newPointers);
+    }
+
+    private MapFetchIndexOperationResultInternal runInternalSorted(
+            InternalIndex index,
+            IndexIterationPointer[] pointers,
+            int fetchLimit
+    ) {
+        List<QueryableEntry> entries = new ArrayList<>();
+        int totalFetched = 0;
+        Comparable lastValueRead = null;
+        boolean done = false;
+
+        IndexIterationPointer[] newPointers = new IndexIterationPointer[pointers.length];
+
+        for (int i = 0; i < pointers.length; i++) {
+            if (done) {
+                newPointers[i] = pointers[i];
+                continue;
+            }
+
+            IndexIterationPointer pointer = pointers[i];
+
+            Iterator<IndexValueBatch> entryIterator = index.getSqlRecordIteratorBatch(
+                    pointer.getFrom(),
+                    pointer.isFromInclusive(),
+                    pointer.getTo(),
+                    pointer.isToInclusive(),
+                    pointer.isDescending()
+            );
+
+            while (entryIterator.hasNext()) {
+                IndexValueBatch indexValueBatch = entryIterator.next();
+                lastValueRead = indexValueBatch.getValue();
+                entries.addAll(indexValueBatch.getEntries());
+                totalFetched += indexValueBatch.getEntries().size();
+
+                if (totalFetched >= fetchLimit) {
+                    done = true;
+                    break;
+                }
+            }
+
+            newPointers[i] = (totalFetched < fetchLimit)
+                    ? null
+                    : new IndexIterationPointer(
+                    pointer.isDescending() ? pointer.getFrom() : lastValueRead,
+                    pointer.isDescending() ? pointer.isFromInclusive() : false,
+                    pointer.isDescending() ? lastValueRead : pointer.getTo(),
+                    pointer.isToInclusive() ? false : pointer.isToInclusive(),
+                    pointer.isDescending());
+
+        }
+
+        return new MapFetchIndexOperationResultInternal(entries, newPointers);
+    }
+
+    private MapFetchIndexOperationResultInternal runInternalHash(
+            InternalIndex index,
+            IndexIterationPointer[] pointers,
+            int fetchLimit
+    ) {
+        IndexIterationPointer[] newPointers = new IndexIterationPointer[pointers.length];
+        List<QueryableEntry> entries = new ArrayList<>();
+        boolean done = false;
+        for (int i = 0; i < pointers.length; i++) {
+            if (done) {
+                newPointers[i] = pointers[i];
+                continue;
+            }
+
+            IndexIterationPointer pointer = pointers[i];
+            entries.addAll(index.getRecords(pointer.getFrom()));
```

Comment:
We can add `assert point.getFrom().equals(pointer.getTo()))` here for the benefit of the reader, because it's not obvious that this must be the case for hash index. Or maybe we can require `getTo()` to return null - to not serialize the value twice in `IndexIterationPointer`.